### PR TITLE
test: migrate fetch and list_availables to snapbox

### PR DESCRIPTION
**Pull request by eth3lbert**
### What does this PR try to resolve?

Part of https://github.com/rust-lang/cargo/issues/14039.

Migrate following to snapbox:

- `tests/testsuite/fetch.rs`
- `tests/testsuite/list_availables.rs`

### CODE REVIEW

File: tests/testsuite/list_availables.rs
Comment (eth3lbert) [inline tests/testsuite/list_availables.rs:1]: Should we make all this snapshot-aware?

Comment (weihanglo) [reply tests/testsuite/list_availables.rs:1]: Yes. That helps if we are going to tweak the error messages.

Comment (eth3lbert) [reply tests/testsuite/list_availables.rs:1]: In order to achieve this goal, I've added a `SnapshotBuilder` to the testsuite. This, along with `ExpectedSnapshots`, also eliminates the need for bit masking in favor of pattern matching.

Comment (weihanglo) [reply tests/testsuite/list_availables.rs:1]: I am not sure whether `SnapshotBuilder` is better, though it isn't really a blocker. If somebody feels unease of it, we can always revert or improve later

**Review by weihanglo — APPROVED** [review]

**Comment by weihanglo** [comment]
@bors r+

Thank you!

**Pull request merged**